[vcpkg] Fix '/' incorrectly replaced by '-' on Linux.

Nemirtingas opened the pull request:
**Describe the pull request**
This PR makes the '/' to be kept on Linux when cross-compiling using a MSVC compatible compiler (like Clang-CL) instead of behind replaced by '-'.

- #### What does your PR fix?
  Fixes #24095

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
Same as before, No.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
Yes.

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
Yes.

Commit:
Fix '/' incorrectly replaced by '-' on Linux.

Commit:
Update baseline.

Commit:
Use MATCHES instead of STREQUAL.

Commit:
Update version hash.

dg0yt commented on ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt:
```diff
@@ -143,7 +143,7 @@ foreach(incdir IN LISTS CMAKE_C_STANDARD_INCLUDE_DIRECTORIES)
 
 foreach(flag CXX C SHARED_LINKER EXE_LINKER STATIC_LINKER RC)
-    if(MSVC)
+    if(MSVC AND "${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Windows")
```
That's an unusual setup. Without an extra comment, the `AND` part will look as if implied by `MSVC` and might get removed in future cleanup.
(IMO if something is claiming to be MSVC, it should really accept the same parameters as `MSVC`.)

(And maybe updates to this PR can wait for a low-load time on CI? We have several changes to vcpkg-cmake now, all building for hours, and affecting more common setups.)

Nemirtingas replied:
So, I should add a comment about cross-compilation so it doesn't get removed by someone else in the future ?
MSVC only implies a MSVC compatible compiler, not that its compiling on Windows.

dg0yt replied:
> MSVC only implies a MSVC compatible compiler, not that its compiling on Windows.

So is your toolchain providing "a `MSVC` compatible compiler" if it is not fully [simulating the Visual C++ cl command-line syntax](https://cmake.org/cmake/help/v3.23/variable/MSVC.html)?

I don't want to say that assuming Windows is right. I'm saying that people might easily accept the assumption that `MSVC` implies `CMAKE_HOST_WIN32`.

Nemirtingas replied:
My toolchain just setup LLVM + Clang + Clang-CL (the CL.exe syntax emulator) for Linux to build Windows binaries on Linux.
As far as I tested, it understands the CL.exe flags/options and transcribes them to Clang valid flags/options (or ignoring them if not needed like /MP (multi-processor compile)). But it also adds some other options that are not provided by CL.exe. Thoses options are rewritten by this rule and destroys all the Toolchain logic.
I have this kind of extra options for clang-cl: (Remember clang-cl is a frontend, it's not a compiler like clang is).
```
-imsvc /clang_windows_sdk/msvc/include
-imsvc /clang_windows_sdk/winsdk/Include/10.0.18362.0/ucrt
-imsvc /clang_windows_sdk/winsdk/Include/10.0.18362.0/shared
-imsvc /clang_windows_sdk/winsdk/Include/10.0.18362.0/um
-imsvc /clang_windows_sdk/winsdk/Include/10.0.18362.0/winrt
```
Which pass the include directories to the clang compiler.

```
-Xclang -ivfsoverlay -Xclang /vcpkg/buildtrees/boost-exception/x86-windows-nemirtingas-rel/winsdk_vfs_overlay.yaml
```
Which pass a virtual FS overlay to clang for case-insensible include files/libraries.

Nemirtingas replied:
The condition I added might not be fixing the issue but only my specific issue. If you run LLVM + Clang + Clang-CL on Windows, the bug should be still there because `"${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Windows"` is true in that case.

strega-nil-ms replied:
I think the right thing to do here is to just add the comment; something like `` # for cross-compilation, we don't want to do this in case we have absolute paths starting with `/` ``

dg0yt commented on ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt:
```diff
@@ -143,7 +143,7 @@ foreach(incdir IN LISTS CMAKE_C_STANDARD_INCLUDE_DIRECTORIES)
+    if(MSVC AND "${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Windows")
         # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
         # This is imperfect because it fails on directories with trailing spaces, but those are exceedingly rare
         string(REGEX REPLACE "(^| )/" "\\1-" ${flag}_FLAGS "${${flag}_FLAGS}")
```
@ras0219-msft I would be curious about the actual cases which motivated this line. Bash shouldn't treat paths different from other arguments, exept that it may do expansion of `*` or  `~`. (In a bash world, you would resolve this with single quotes.)
But MSYS2 has special treatment for converting Unix paths to Windows paths in command line arguments and environment variables. This could probably be resolved by setting enviroment variables to declare the known flags.

Neumann-A replied:
> bash scripts

Means libtool and other stuff run via configure. Just try to remove it and see the `vcpkg_configure_make` world burn ;)

See #12936. Especially https://github.com/microsoft/vcpkg/pull/12936/commits/65e71b182261d0b61dda1fcef23083ec1f5eb4fc (I aborted the CI run since I already saw it failing) vs https://github.com/microsoft/vcpkg/pull/12936/commits/723b9d77dbd3051b8769c251310eda483b71a5d5

dg0yt replied:
> Means libtool and other stuff run via configure.

Means MSYS2 runtime, not bash.

dg0yt replied:
Illustration
~~~sh
$ /usr/bin/echo.exe /libpath/users
/libpath/users
$ /C/Windows/System32/findstr.exe a /libpath/users
FINDSTR: D:/msys64/libpath/users kann nicht geöffnet werden.
$ MSYS2_ARG_CONV_EXCL=/libpath /C/Windows/System32/findstr.exe a /libpath/users
FINDSTR: /libpath/users kann nicht geöffnet werden.
~~~
I see that the `MSYS2_ARG_CONV_EXCL` doesn't scale to all combinations of tools and parameters, even if scoped via wrappers.
But sometimes it is necessary to understand that the transformation is not in `bash` itself (cf. bash -> echo) but a MSYS2 service at the border between MSYS2 runtime (bash) and windows runtime (findstr).

Neumann-A replied:
> Means MSYS2 runtime, not bash.

libtool is just a bash script. But I really don't care about the details anymore. vcpkg should just build the msys stuff all by itself and no longer have to deal with MSYS2 since it has often enough shown to be instable by removing downloadable stuff.

Commit:
Add comment for future modifications.

Commit:
Update version hash.

Nemirtingas commented:
> I think the right thing to do here is to just add the comment; something like `` # for cross-compilation, we don't want to do this in case we have absolute paths starting with `/` `` (otherwise I'm happy with this change, personally)

I'm happy with the change too. I added the comment and updated the version hash.

strega-nil-ms commented:
/azp run

LilyWangLL commented:
Depends on #24169

Nemirtingas commented:
> Depends on #24169

Why does this PR depends on #24169 ?

LilyWangLL commented:
> > Depends on #24169
> 
> Why does this PR depends on #24169 ?

This PR mark `zeroc-ice` doesn't support `staticrt` on windows, it should fix the pipeline failed.

Nemirtingas commented:
> > > Depends on #24169
> > 
> > 
> > Why does this PR depends on #24169 ?
> 
> This PR mark `zeroc-ice` doesn't support `staticrt` on windows, it should fix the pipeline failed.

Oh, ok, that was about the pipeline :).

Commit:
Merge branch 'master' into fix_slash_replaced_by_dash

Commit:
Update version-date.

Commit:
Update baseline.

BillyONeal commented on ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt:
```diff
@@ -143,7 +143,8 @@ foreach(incdir IN LISTS CMAKE_C_STANDARD_INCLUDE_DIRECTORIES)
 
 foreach(flag CXX C SHARED_LINKER EXE_LINKER STATIC_LINKER RC)
-    if(MSVC)
+    # For cross-compilation, we don't want to do this in case we have absolute paths starting with `/`
+    if(MSVC AND "${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Windows")
         # Transform MSVC /flags to -flags due to bash scripts intepreting /flag as a path.
         # This is imperfect because it fails on directories with trailing spaces, but those are exceedingly rare
```
```suggestion
    # When using MSVC, maybe transform /flags to -flags.
    #   When cross compiling, "/flags" may be an absolute path starting with /, so don't transform.
    #   Otherwise, we do this transformation due to scripts interpreting "/flags" as a path.
    if(MSVC AND "${CMAKE_HOST_SYSTEM_NAME}" MATCHES "Windows")
        # This implementation is imperfect because it fails on directories with trailing spaces,
        # but those are rare.
```

BillyONeal reviewed (approved):
This is OK to me assuming CMAKE_HOST_SYSTEM_NAME contains Windows under all our "desktop" and "uwp" triplets and does not under MinGW.

LilyWangLL commented:
Ping @Nemirtingas for response, could you resolve the conflicts?

Nemirtingas committed:
Update ports/vcpkg-cmake/cmake_get_vars/CMakeLists.txt

Co-authored-by: Billy O'Neal <bion@microsoft.com>

Commit:
Merge remote-tracking branch 'vcpkg/master' into fix_slash_replaced_by_dash

Commit:
Update version.

Commit:
Run vcpkg format-manifest ports/vcpkg-cmake/vcpkg.json

Commit:
Run vcpkg x-add-version vcpkg-cmake

Nemirtingas commented:
> Ping @Nemirtingas for response, could you resolve the conflicts?

Done.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

# Conflicts:
#	versions/v-/vcpkg-cmake.json

BillyONeal reviewed (approved):
Slash dot dash dot slash dot dash dot slash dot com

BillyONeal commented:
I merged this despite not having a full build for it since there were no 'interesting' merge conflicts outside of the version database.

Thanks for the fix!
